### PR TITLE
Stopcock Rate limit

### PR DESCRIPTION
**Pull request by dhavalsoni9989**
- feature: Added the stopcock library for throttled the API limit

### CODE REVIEW

File: src/KnawatMP.js [hunk 1]
```diff
@@ -1,8 +1,16 @@
 import fetch from 'node-fetch';
 import qs from 'qs';
+import stopcock from 'stopcock';
 
 class KnawatMP {
   static baseUrl = process.env.KNAWAT_MP_BASE_URL || 'https://mp.knawat.io/api';
+
+  static stopcockRequest = stopcock(fetch, {
```
Comment (HazemKhaled) [inline src/KnawatMP.js:8]: You forget to make `npm i stopcock --save`

File: src/KnawatMP.js [hunk 2]
```diff
@@ -1,8 +1,16 @@
...
+    bucketSize: process.env.STOPCOCK_BUCKET_SIZE || 120,
```
Comment (HazemKhaled) [inline src/KnawatMP.js:9]: use these number please
https://docs.knawat.io/#section/Knawat-API-rate-limit

**Review by HazemKhaled — CHANGES_REQUESTED** [review]

File: src/KnawatMP.js (hunk 1 above)
Comment (dhavalsoni9989) [reply src/KnawatMP.js:8]: @HazemKhaled 

I have added the stopcock in package.json file

File: src/KnawatMP.js (hunk 2 above)
Comment (dhavalsoni9989) [reply src/KnawatMP.js]: updated the code

**Comment by ismail9k** [comment]
Form API using prospective, I think configuring the stopcock thought env variable is not the optimal solution, because they should not be constant across the app (unlike the base URL for example).
My proposed API using is passing the configs as the second argument, and giving the ability the disable throttling.

Passing no configs will be enabled by default
```js
 const mp = new MP({  store });
```

Updating only on value, the other values should go the defaults
```js
 const mp = new MP({  store }, { autoLimit: { bucketSize: 12 } });
```

Disable autoLimit
```js
 const mp = new MP({  store }, { autoLimit: false });
```

> P.S please update the readme file accordingly

File: src/KnawatMP.js [hunk 3]
```diff
@@ -1,8 +1,16 @@
 import fetch from 'node-fetch';
 import qs from 'qs';
+import stopcock from 'stopcock';
 
 class KnawatMP {
   static baseUrl = process.env.KNAWAT_MP_BASE_URL || 'https://mp.knawat.io/api';
+
+  static stopcockRequest = stopcock(fetch, {
+    bucketSize: process.env.STOPCOCK_BUCKET_SIZE || 2,
+    interval: process.env.STOPCOCK_INTERVAL || 2000,
```
Comment (HazemKhaled) [inline src/KnawatMP.js:10]: According to docs it should be 2/second
https://docs.knawat.io/#section/Knawat-API-rate-limit

File: src/KnawatMP.js [hunk 4]
```diff
@@ -1,8 +1,16 @@
 import fetch from 'node-fetch';
 import qs from 'qs';
+import stopcock from 'stopcock';
 
 class KnawatMP {
   static baseUrl = process.env.KNAWAT_MP_BASE_URL || 'https://mp.knawat.io/api';
+
+  static stopcockRequest = stopcock(fetch, {
+    bucketSize: process.env.STOPCOCK_BUCKET_SIZE || 2,
```
Comment (HazemKhaled) [inline src/KnawatMP.js:9]: bucketSize = 2! according to the Knawat docs, we don't watch the bucket size, so please update the docs or update the code here

**Review by HazemKhaled — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -44,15 +44,46 @@ Setup for the new Knawat Dropshipping REST API integration:
 const MP = require('@knawat/mp');
 
 // Provide instance with your store credentials
+// If want to use the stopcock the pass the autoLimit object
 const mp = new MP({
   key: 'XXXXXXXXXXXXXXXXXXXXXXXXXXX',
   secret: 'XXXXXXXXXXXXXXXXXXXXXXXX',
+  autoLimit: { bucketSize: 2, interval: 1000, limit: 1 },
```
Comment (HazemKhaled) [inline README.md:51]: autoLimit is optional, don't update all examples, just add one example in case we need to overwrite the values

**Review by HazemKhaled — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (dhavalsoni9989) [reply README.md:51]: Updated the code

File: src/KnawatMP.js (hunk 4 above)
Comment (dhavalsoni9989) [reply src/KnawatMP.js]: Added the unit test cases and updated the code

File: src/KnawatMP.js [hunk 6]
```diff
@@ -20,10 +24,11 @@ class KnawatMP {
    * KnawatMP sdk constructor
    * @param {*} param0 {key: string|null, secret: string|null, store: string|null}
    */
-  constructor({ key, secret, store } = {}) {
+  constructor({ key, secret, store, autoLimit } = {}) {
     this.key = key;
     this.secret = secret;
     this.store = store;
+    this.autoLimit = autoLimit || {};
```
Comment (ismail9k) [inline src/KnawatMP.js:31]: This will break if I passed `autoLimit: false`

File: src/KnawatMP.js [hunk 7]
```diff
@@ -46,6 +51,13 @@ class KnawatMP {
    * Get the current store credentials
    */
   async setCurrentStoreCredentials() {
+    if (this.autoLimit) {
```
Comment (ismail9k) [inline src/KnawatMP.js:54]: `autoLimit` will always be true

File: src/KnawatMP.js (hunk 6 above)
Comment (HazemKhaled) [reply src/KnawatMP.js:31]: You shouldn't disable it, MP will fail your requests if you ignore it

Comment (HazemKhaled) [reply src/KnawatMP.js:31]: But you can create an instance with apiRateLimit.limit = 99

but you will get this error ;)
image

**Review by ismail9k — APPROVED** [review]

**Pull request merged**